File: apps/backend/src/modules/maintenance-ticket/entities/maintenance-ticket.entity.ts
import {
  Entity,
  Column,
  Index,
  ManyToOne,
  OneToMany,
  JoinColumn,
} from 'typeorm';
import { TenantBaseEntity } from '../../../shared/database/tenant-base.entity';
import { TicketStatus } from '../enums/ticket-status.enum';
import { TicketPriority } from '../enums/ticket-priority.enum';
import { TicketType } from '../enums/ticket-type.enum';
import { TicketStatusHistory } from './ticket-status-history.entity';
import { TicketComment } from './ticket-comment.entity';
import { TicketAttachment } from './ticket-attachment.entity';
import { Department } from './department.entity';
import { TicketCategory } from './ticket-category.entity';
import { Team } from './team.entity';
import { User } from '../../iam/entities/user.entity';
import { Site } from '../../tenant/entities/site.entity';
import { Space } from '../../tenant/entities/space.entity';
import { Villa } from '../../tenant/entities/villa.entity';

@Entity('maintenance_tickets')
// Identification indexes
@Index(['companyId', 'ticketNumber'], { unique: true })
@Index(['companyId', 'ticketType'])
// Location indexes
@Index(['companyId', 'villaNumber'])
@Index(['companyId', 'villaId'])
@Index(['companyId', 'siteId'])
@Index(['companyId', 'spaceId'])
// Status/filtering indexes
@Index(['companyId', 'status'])
@Index(['companyId', 'priority'])
@Index(['companyId', 'createdBy'])
// Assignment indexes
@Index(['companyId', 'assignedTechnicianId'])
@Index(['companyId', 'assignedSupervisorId'])
@Index(['companyId', 'assignedTeamId'])
@Index(['companyId', 'categoryId'])
// Hierarchy index
@Index(['companyId', 'parentTicketId'])
export class MaintenanceTicket extends TenantBaseEntity {
  @Column({ type: 'varchar', length: 50, name: 'ticket_number' })
  ticketNumber!: string;

  // ============================================================================
  // TICKET TYPE (NEW - from service_requests)
  // ============================================================================
  @Column({
    type: 'enum',
    enum: TicketType,
    default: TicketType.MAINTENANCE,
    name: 'ticket_type',
  })
  ticketType!: TicketType;

  // ============================================================================
  // LOCATION (villa_number kept for backward compatibility, villa_id is preferred)
  // ============================================================================
  /**
   * @deprecated Use villaId instead. Kept for backward compatibility.
   */
  @Column({ type: 'varchar', length: 50, name: 'villa_number', nullable: true })
  villaNumber?: string;

  @Column({ type: 'uuid', name: 'villa_id', nullable: true })
  villaId?: string;

  @ManyToOne(() => Villa, { nullable: true })
  @JoinColumn({ name: 'villa_id' })
  villa?: Villa;

  @Column({ type: 'uuid', name: 'site_id', nullable: true })
  siteId?: string;

  @ManyToOne(() => Site, { nullable: true })
  @JoinColumn({ name: 'site_id' })
  site?: Site;

  @Column({ type: 'uuid', name: 'space_id', nullable: true })
  spaceId?: string;

  @ManyToOne(() => Space, { nullable: true })
  @JoinColumn({ name: 'space_id' })
  space?: Space;

  @Column({ type: 'uuid', name: 'created_by', nullable: false })
  createdBy!: string;

  @ManyToOne(() => User, { nullable: false })
  @JoinColumn({ name: 'created_by' })
  creator!: User;

  @Column({ type: 'varchar', length: 255 })
  title!: string;

  @Column({ type: 'text', nullable: true })
  description?: string;

  /**
   * Free-text location detail inside the villa/space
   * Examples: "Kitchen", "Living Room", "Bedroom 2"
   */
  @Column({
    type: 'varchar',
    length: 255,
    name: 'location_detail',
    nullable: true,
  })
  locationDetail?: string;

  @Column({
    type: 'enum',
    enum: TicketStatus,
    default: TicketStatus.NEW,
  })
  status!: TicketStatus;

  @Column({
    type: 'enum',
    enum: TicketPriority,
    default: TicketPriority.MEDIUM,
  })
  priority!: TicketPriority;

  @Column({ type: 'uuid', name: 'category_id', nullable: true })
  categoryId?: string;

  @ManyToOne(() => TicketCategory, { nullable: true })
  @JoinColumn({ name: 'category_id' })
  category?: TicketCategory;

  // ============================================================================
  // TENANT CONTACT INFORMATION
  // ============================================================================
  @Column({ type: 'varchar', length: 50, name: 'contact_number', nullable: true })
  contactNumber?: string;

  @Column({ type: 'varchar', length: 50, name: 'alternate_contact', nullable: true })
  alternateContact?: string;

  @Column({ type: 'varchar', length: 255, name: 'preferred_time', nullable: true })
  preferredTime?: string;

  @Column({ type: 'uuid', name: 'department_id', nullable: true })
  departmentId?: string;

  @ManyToOne(() => Department, { nullable: true })
  @JoinColumn({ name: 'department_id' })
  department?: Department;

  @Column({ type: 'uuid', name: 'assigned_supervisor_id', nullable: true })
  assignedSupervisorId?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: 'assigned_supervisor_id' })
  assignedSupervisor?: User;

  @Column({
    type: 'timestamp with time zone',
    name: 'supervisor_assigned_at',
    nullable: true,
  })
  supervisorAssignedAt?: Date;

  @Column({ type: 'uuid', name: 'assigned_technician_id', nullable: true })
  assignedTechnicianId?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: 'assigned_technician_id' })
  assignedTechnician?: User;

  @Column({ type: 'uuid', name: 'assigned_by', nullable: true })
  assignedBy?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: 'assigned_by' })
  assigner?: User;

  @Column({
    type: 'timestamp with time zone',
    name: 'assigned_at',
    nullable: true,
  })
  assignedAt?: Date;

  @Column({ type: 'uuid', name: 'acknowledged_by', nullable: true })
  acknowledgedBy?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: 'acknowledged_by' })
  acknowledger?: User;

  @Column({
    type: 'timestamp with time zone',
    name: 'acknowledged_at',
    nullable: true,
  })
  acknowledgedAt?: Date;

  @Column({
    type: 'timestamp with time zone',
    name: 'scheduled_at',
    nullable: true,
  })
  scheduledAt?: Date;

  @Column({ type: 'text', nullable: true, name: 'technician_notes' })
  technicianNotes?: string;

  @Column({ type: 'text', nullable: true, name: 'resolution_notes' })
  resolutionNotes?: string;

  @Column({
    type: 'timestamp with time zone',
    name: 'completed_at',
    nullable: true,
  })
  completedAt?: Date;

  @Column({
    type: 'timestamp with time zone',
    name: 'closed_at',
    nullable: true,
  })
  closedAt?: Date;

  @Column({
    type: 'timestamp with time zone',
    name: 'auto_close_at',
    nullable: true,
  })
  autoCloseAt?: Date;

  @Column({ type: 'boolean', default: false, name: 'tenant_confirmed' })
  tenantConfirmed!: boolean;

  // ============================================================================
  // RATING (NEW - tenant rating after ticket closure)
  // ============================================================================
  @Column({ type: 'integer', nullable: true })
  rating?: number;

  @Column({ type: 'text', name: 'rating_comment', nullable: true })
  ratingComment?: string;

  @Column({
    type: 'timestamp with time zone',
    name: 'rated_at',
    nullable: true,
  })
  ratedAt?: Date;

  @Column({ type: 'uuid', name: 'rated_by', nullable: true })
  ratedBy?: string;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: 'rated_by' })
  ratedByUser?: User;

  // ============================================================================
  // TEAM ASSIGNMENT (NEW - from service_requests)
  // ============================================================================
  @Column({ type: 'uuid', name: 'assigned_team_id', nullable: true })
  assignedTeamId?: string;

  @ManyToOne(() => Team, { nullable: true })
  @JoinColumn({ name: 'assigned_team_id' })
  assignedTeam?: Team;

  // ============================================================================
  // PARENT/CHILD TICKETS (NEW - from service_requests)
  // ============================================================================
  @Column({ type: 'uuid', name: 'parent_ticket_id', nullable: true })
  parentTicketId?: string;

  @ManyToOne(() => MaintenanceTicket, (ticket) => ticket.childTickets, { nullable: true })
  @JoinColumn({ name: 'parent_ticket_id' })
  parentTicket?: MaintenanceTicket;

  @OneToMany(() => MaintenanceTicket, (ticket) => ticket.parentTicket)
  childTickets!: MaintenanceTicket[];

  // ============================================================================
  // ESCALATION (NEW - from service_requests)
  // ============================================================================
  @Column({ type: 'boolean', default: false, name: 'is_escalated' })
  isEscalated!: boolean;

  @Column({ type: 'int', default: 0, nullable: true, name: 'escalation_level' })
  escalationLevel?: number;

  @Column({
    type: 'timestamp with time zone',
    name: 'escalated_at',
    nullable: true,
  })
  escalatedAt?: Date;

  // ============================================================================
  // RELATIONS
  // ============================================================================
  @OneToMany(() => TicketStatusHistory, (history) => history.ticket, {
    cascade: true,
  })
  statusHistory!: TicketStatusHistory[];

  @OneToMany(() => TicketComment, (comment) => comment.ticket, {
    cascade: true,
  })
  comments!: TicketComment[];

  @OneToMany(() => TicketAttachment, (attachment) => attachment.ticket, {
    cascade: true,
  })
  attachments!: TicketAttachment[];
}

File: apps/frontend/lib/src/features/maintenance_ticket/domain/entities/maintenance_ticket_entity.dart
import 'package:equatable/equatable.dart';

import '../../../auth/domain/entities/user_entity.dart';
import 'ticket_type_entity.dart';
import 'villa_entity.dart';
import 'team_entity.dart';
import 'site_entity.dart';
import 'space_entity.dart';
import 'ticket_category_entity.dart';

enum TicketStatus {
  new_,
  acknowledged,
  assigned,
  inProgress,
  onHold,
  completed,
  cancelled;

  static TicketStatus fromString(String value) {
    // Normalize both the input and enum names for comparison
    final normalizedValue = value.toLowerCase().replaceAll('_', '');
    return TicketStatus.values.firstWhere(
      (e) => e.name.toLowerCase() == normalizedValue,
      orElse: () => TicketStatus.new_,
    );
  }

  String get displayName {
    switch (this) {
      case TicketStatus.new_:
        return 'New';
      case TicketStatus.acknowledged:
        return 'Acknowledged';
      case TicketStatus.assigned:
        return 'Assigned';
      case TicketStatus.inProgress:
        return 'In Progress';
      case TicketStatus.onHold:
        return 'On Hold';
      case TicketStatus.completed:
        return 'Completed';
      case TicketStatus.cancelled:
        return 'Cancelled';
    }
  }

  /// Converts the enum to the backend API format (uppercase with underscores)
  String get toBackendValue {
    switch (this) {
      case TicketStatus.new_:
        return 'NEW';
      case TicketStatus.acknowledged:
        return 'ACKNOWLEDGED';
      case TicketStatus.assigned:
        return 'ASSIGNED';
      case TicketStatus.inProgress:
        return 'IN_PROGRESS';
      case TicketStatus.onHold:
        return 'ON_HOLD';
      case TicketStatus.completed:
        return 'COMPLETED';
      case TicketStatus.cancelled:
        return 'CANCELLED';
    }
  }
}

enum TicketPriority {
  low,
  medium,
  high,
  urgent;

  static TicketPriority fromString(String value) {
    return TicketPriority.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => TicketPriority.medium,
    );
  }

  String get displayName {
    return name.toUpperCase();
  }

  /// Converts the enum to the backend API format (uppercase)
  String get toBackendValue {
    return name.toUpperCase();
  }
}

class MaintenanceTicketEntity extends Equatable {
  const MaintenanceTicketEntity({
    required this.id,
    this.companyId,
    required this.ticketNumber,
    this.ticketType,
    this.villaNumber,
    this.villaId,
    this.villa,
    this.siteId,
    this.site,
    this.spaceId,
    this.space,
    required this.createdBy,
    this.creator,
    this.creatorName,
    required this.title,
    this.description,
    this.locationDetail,
    this.contactNumber,
    this.alternateContact,
    this.preferredTime,
    required this.status,
    required this.priority,
    this.priorityDetails,
    // Classification
    this.categoryId,
    this.category,
    this.departmentId,
    this.department,
    this.departmentName,
    // Assignment fields
    this.assignedSupervisorId,
    this.assignedSupervisor,
    this.assignedSupervisorName,
    this.supervisorAssignedAt,
    this.assignedTechnicianId,
    this.assignedTechnician,
    this.assignedTechnicianName,
    this.assignedTeamId,
    this.assignedTeam,
    this.assignedBy,
    this.assigner,
    this.assignedAt,
    this.acknowledgedBy,
    this.acknowledger,
    this.acknowledgedAt,
    this.scheduledAt,
    this.technicianNotes,
    this.resolutionNotes,
    this.completedAt,
    this.closedAt,
    this.autoCloseAt,
    required this.tenantConfirmed,
    // Rating fields
    this.rating,
    this.ratingComment,
    this.ratedAt,
    this.ratedBy,
    // Parent/Child relationships
    this.parentTicketId,
    // Escalation fields
    this.isEscalated = false,
    this.escalationLevel = 0,
    this.escalatedAt,
    // SLA fields
    this.slaDueAt,
    this.slaStatus,
    // Timestamps
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String? companyId;
  final String ticketNumber;
  final TicketType? ticketType;

  // Location fields
  final String? villaNumber;
  final String? villaId;
  final VillaEntity? villa;
  final String? siteId;
  final SiteEntity? site;
  final String? spaceId;
  final SpaceEntity? space;

  // Creator
  final String createdBy;
  final UserEntity? creator;
  final String? creatorName;

  // Content
  final String title;
  final String? description;
  final String? locationDetail;
  final String? contactNumber;
  final String? alternateContact;
  final String? preferredTime;
  final TicketStatus status;
  final TicketPriority priority;
  final PriorityDetailsEntity? priorityDetails;

  // Classification
  final String? categoryId;
  final TicketCategoryEntity? category;
  final String? departmentId;
  final DepartmentEntity? department;
  final String? departmentName;

  // Assignment
  final String? assignedSupervisorId;
  final UserEntity? assignedSupervisor;
  final String? assignedSupervisorName;
  final DateTime? supervisorAssignedAt;
  final String? assignedTechnicianId;
  final UserEntity? assignedTechnician;
  final String? assignedTechnicianName;
  final String? assignedTeamId;
  final TeamEntity? assignedTeam;
  final String? assignedBy;
  final UserEntity? assigner;
  final DateTime? assignedAt;
  final String? acknowledgedBy;
  final UserEntity? acknowledger;
  final DateTime? acknowledgedAt;
  final DateTime? scheduledAt;
  final String? technicianNotes;
  final String? resolutionNotes;
  final DateTime? completedAt;
  final DateTime? closedAt;
  final DateTime? autoCloseAt;
  final bool tenantConfirmed;

  // Rating fields
  final int? rating;
  final String? ratingComment;
  final DateTime? ratedAt;
  final String? ratedBy;

  // Parent/Child relationships
  final String? parentTicketId;

  // Escalation
  final bool isEscalated;
  final int escalationLevel;
  final DateTime? escalatedAt;

  // SLA Information
  final DateTime? slaDueAt;
  final String? slaStatus;

  // Timestamps
  final DateTime createdAt;
  final DateTime updatedAt;

  /// Helper: Get location display string
  String get locationDisplay {
    if (villa != null) {
      return villa!.displayName;
    }
    if (site != null && space != null) {
      return '${site!.displayName} → ${space!.displayName}';
    }
    if (site != null) {
      return site!.displayName;
    }
    if (space != null) {
      return space!.displayName;
    }
    // Fallback to deprecated villa_number for backward compatibility
    if (villaNumber != null) {
      return 'Villa $villaNumber';
    }
    return 'Unknown Location';
  }

  /// Helper: Check if ticket has parent
  bool hasParent() => parentTicketId != null && parentTicketId!.isNotEmpty;

  /// Helper: Check if ticket is escalated
  bool isEscalatedTicket() => isEscalated && escalationLevel > 0;

  /// Helper: Get ticket type display name
  String get ticketTypeDisplay {
    return ticketType?.displayName ?? 'Unknown';
  }

  /// Helper: Get ticket type color
  int? get ticketTypeColor => ticketType?.colorValue;

  /// Helper: Get creator name (from entity or fallback to stored name)
  String? get creatorDisplayName {
    if (creator != null) {
      final name =
          '${creator!.firstName ?? ''} ${creator!.lastName ?? ''}'.trim();
      return name.isNotEmpty ? name : creator!.email;
    }
    return creatorName;
  }

  /// Helper: Get department name (from entity or fallback to stored name)
  String? get departmentDisplayName {
    return department?.name ?? departmentName;
  }

  /// Helper: Get assigned supervisor name (from entity or fallback to stored name)
  String? get assignedSupervisorDisplayName {
    if (assignedSupervisor != null) {
      final name =
          '${assignedSupervisor!.firstName ?? ''} ${assignedSupervisor!.lastName ?? ''}'
              .trim();
      return name.isNotEmpty ? name : assignedSupervisor!.email;
    }
    return assignedSupervisorName;
  }

  /// Helper: Get assigned technician name (from entity or fallback to stored name)
  String? get assignedTechnicianDisplayName {
    if (assignedTechnician != null) {
      final name =
          '${assignedTechnician!.firstName ?? ''} ${assignedTechnician!.lastName ?? ''}'
              .trim();
      return name.isNotEmpty ? name : assignedTechnician!.email;
    }
    return assignedTechnicianName;
  }

  @override
  List<Object?> get props => [
        id,
        companyId,
        ticketNumber,
        ticketType,
        villaNumber,
        villaId,
        villa,
        siteId,
        site,
        spaceId,
        space,
        createdBy,
        creator,
        creatorName,
        title,
        description,
        locationDetail,
        contactNumber,
        alternateContact,
        preferredTime,
        status,
        priority,
        priorityDetails,
        categoryId,
        category,
        departmentId,
        department,
        departmentName,
        assignedSupervisorId,
        assignedSupervisor,
        assignedSupervisorName,
        supervisorAssignedAt,
        assignedTechnicianId,
        assignedTechnician,
        assignedTechnicianName,
        assignedTeamId,
        assignedTeam,
        assignedBy,
        assigner,
        assignedAt,
        acknowledgedBy,
        acknowledger,
        acknowledgedAt,
        scheduledAt,
        technicianNotes,
        resolutionNotes,
        completedAt,
        closedAt,
        autoCloseAt,
        tenantConfirmed,
        rating,
        ratingComment,
        ratedAt,
        ratedBy,
        parentTicketId,
        isEscalated,
        escalationLevel,
        escalatedAt,
        slaDueAt,
        slaStatus,
        createdAt,
        updatedAt,
      ];
}

class PriorityDetailsEntity extends Equatable {
  const PriorityDetailsEntity({
    this.colorCode,
    this.iconName,
    this.defaultSlaHours,
  });

  final String? colorCode;
  final String? iconName;
  final int? defaultSlaHours;

  @override
  List<Object?> get props => [colorCode, iconName, defaultSlaHours];
}

class DepartmentEntity extends Equatable {
  const DepartmentEntity({
    required this.id,
    required this.name,
    this.description,
    required this.isActive,
  });

  final String id;
  final String name;
  final String? description;
  final bool isActive;

  @override
  List<Object?> get props => [id, name, description, isActive];
}
